src/lib.rs

```diff
@@ -23,18 +29,132 @@ mod tests {
     // test if mapping has been added
     #[test]
     fn test_add_mapping() {
-        let addr_space = AddressSpace::new("Test address space");
+        let mut addr_space = AddressSpace::new("Test address space");
         let data_source: FileDataSource = FileDataSource::new("Cargo.toml").unwrap();
         let offset: usize = 0;
         let length: usize = 1;
+        let read_flags = FlagBuilder::new().toggle_read();
 
-        let addr = addr_space.add_mapping(&data_source, offset, length).unwrap();
+        let ds_arc = Arc::new(data_source);
+
+        let addr = addr_space
+            .add_mapping(ds_arc.clone(), offset, length, read_flags)
+            .unwrap();
         assert!(addr != 0);
 
-        // we should move these tests into addr_space, since they access non-public internals of the structure:
-        // assert_eq!(addr_space.mappings.is_empty(), false);
-        // assert_eq!(addr_space.mappings.front().source, Some(&data_source));
-        // assert_eq!(addr_space.mappings.front().offset, offset);
-        // assert_eq!(addr_space.mappings.front().span, length);
+        let addr2 = addr_space
+            .add_mapping(ds_arc.clone(), address_space::PAGE_SIZE, 0, read_flags)
+            .unwrap();
+        assert!(addr2 != 0);
+        assert!(addr2 != addr);
+    }
+
+    // test running out of address space
+    #[test]
+    #[should_panic]
+    fn test_add_mapping_error() {
+        let mut addr_space = AddressSpace::new("Test address space for errors");
+        let data_source: FileDataSource = FileDataSource::new("Cargo.toml").unwrap();
+        let offset: usize = 0;
+        let length: usize = 1 << 38;
+        let read_flags = FlagBuilder::new().toggle_read();
+
+        let ds_arc = Arc::new(data_source);
+
+        let addr = addr_space
+            .add_mapping(ds_arc.clone(), offset, length, read_flags)
+            .unwrap();
+    }
+
+    // test if mapping has been added
+    #[test]
+    fn test_add_mapping_at() {
+        let mut addr_space = AddressSpace::new("Test address space");
+        let data_source: FileDataSource = FileDataSource::new("Cargo.toml").unwrap();
+        let offset: usize = 0;
+        let length: usize = 1;
+        let start: usize = 100;
+        let read_flags = FlagBuilder::new().toggle_read();
+
+        let ds_arc = Arc::new(data_source);
+
+        let addr = addr_space
+            .add_mapping_at(ds_arc.clone(), offset, length, start, read_flags)
+            .unwrap();
```

Comment:
Here, I guess you just test that `add_mapping_at` does not return an error. First, that should be done with `assert!(addr.is_ok())`, and second, it might be useful to test that the internals of the address space have been adjusted correctly. You'd need to move the test to `src/address_space.rs` to do that.